### PR TITLE
Fix refresh iterator rate limit

### PR DESCRIPTION
**Pull request by twang-rs**
Also fix data races by using atomic counters + use getErrCount() function instead of directly accessing errCount.

### CODE REVIEW

File: listener.go [hunk 1]
```diff
@@ -208,10 +208,25 @@ func (l *Listener) consume() {
 
 	counter := 0
 	timer := time.Now()
+	needsRefreshIterator := false
 
 	for {
+		if !l.shouldConsume() {
```
Comment (twang-rs) [inline listener.go]: Moved up, so that `initShardIterator` cannot starve a SIG_INT

File: listener.go [hunk 2]
```diff
@@ -208,10 +208,25 @@ func (l *Listener) consume() {
...
+			l.setConsuming(false)
+			break
+		}
+
 		l.throttle(&counter, &timer)
 
+		if needsRefreshIterator {
```
Comment (twang-rs) [inline listener.go]: Move this inline with the loop (and `throttle`), in hopes of reducing the number of `ProvisionedThroughputExceededExceptions`

File: listener_test.go [hunk 3]
```diff
@@ -49,7 +49,7 @@ func TestListenerError(t *testing.T) {
 			// Let the error propagate
 			<-time.After(1 * time.Second)
 
-			So(listener.errCount, ShouldNotEqual, 0)
+			So(listener.getErrCount(), ShouldNotEqual, 0)
```
Comment (twang-rs) [inline listener_test.go]: This was causing a data race (since it was not taking the lock to access the counter).

File: producer_test.go [hunk 4]
```diff
@@ -44,7 +44,7 @@ func TestProducerError(t *testing.T) {
 			producer.errors <- errors.New("All your base are belong to us!")
 			// Let the error propagate
 			<-time.After(1 * time.Second)
-			So(producer.errCount, ShouldEqual, 1)
+			So(producer.getErrCount(), ShouldEqual, 1)
```
Comment (twang-rs) [inline producer_test.go]: This was causing a data race.

File: listener_test.go (hunk 3 above)
Comment (fjordan) [inline listener_test.go]: Thank you for fixing this.

File: firehose.go [hunk 5]
```diff
@@ -143,6 +143,6 @@ func (p *Producer) sendFirehoseRecords(args *gokinesis.RequestArgs) {
 	}
 
 	if conf.Debug.Verbose && p.getMsgCount()%100 == 0 {
-		log.Println("Messages sent so far: " + strconv.Itoa(p.getMsgCount()))
+		log.Println("Messages sent so far: " + strconv.FormatInt(p.getMsgCount(), 10))
```
Comment (fjordan) [inline firehose.go]: From the docs `Itoa` appears to be shorthand for `Itoa is shorthand for FormatInt(int64(i), 10)`. Do we need the longhand here?

Comment (twang-rs) [inline firehose.go]: `Itoa` only works for `int` types.  To use it for int64 (which is what an atomic counter uses), we must use longhand.

Comment (twang-rs) [inline firehose.go]: We could optionally force the `int64` back to `int` in `getErrCount` and `getMsgCount` but I'm not sure if we will get precision loss errors/warnings.

Comment (fjordan) [inline firehose.go]: I see now that the counter was changed to `int64` for the `atomic` operations.

**Pull request merged**